Compilation errors with OAuth and ManyToMany with User on a microservice app

##### **Overview of the issue**

Compilation errors below:

```
[ERROR] /Users/theo/Documents/perso/jhipster-many/mono/src/main/java/com/mycompany/myapp/web/rest/CountryResource.java:[57,20] cannot find symbol
  symbol:   method getUser()
  location: variable country of type com.mycompany.myapp.domain.Country
[ERROR] /Users/theo/Documents/perso/jhipster-many/mono/src/main/java/com/mycompany/myapp/web/rest/CountryResource.java:[59,40] cannot find symbol
  symbol:   method getUser()
  location: variable country of type com.mycompany.myapp.domain.Country
[ERROR] /Users/theo/Documents/perso/jhipster-many/mono/src/main/java/com/mycompany/myapp/web/rest/CountryResource.java:[82,20] cannot find symbol
  symbol:   method getUser()
  location: variable country of type com.mycompany.myapp.domain.Country
[ERROR] /Users/theo/Documents/perso/jhipster-many/mono/src/main/java/com/mycompany/myapp/web/rest/CountryResource.java:[84,40] cannot find symbol
  symbol:   method getUser()
  location: variable country of type com.mycompany.myapp.domain.Country
```

##### **Motivation for or Use Case**

Project should compile.

##### **Reproduce the error**

Generate the app using the JDL below:

<details>
<summary>JDL file</summary>
<pre>
application {
  config {
    authenticationType oauth2
    applicationType microservice
  }
  entities *
}
entity Country {
  name String
}
relationship ManyToMany {
  Country{User(login)} to User{Country(name)}
}
</pre>
</details>

##### **Related issues**

#11160

##### **Suggest a Fix**

<!-- If you can't fix the bug yourself, perhaps you can point to what might be
  causing the problem (line of code or commit) -->

##### **JHipster Version(s)**

```
jhipster@0.0.0 /Users/theo/Documents/perso/jhipster-many/mono
└── generator-jhipster@6.6.0  -> /Users/theo/Documents/perso/generator-jhipster-fork

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": ["en", "fr"],
    "messageBroker": false,
    "nativeLanguage": "en",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "serviceDiscoveryType": "eureka",
    "skipClient": true,
    "testFrameworks": [],
    "websocket": false,
    "authenticationType": "oauth2",
    "applicationType": "microservice",
    "jhipsterVersion": "6.6.0",
    "creationTimestamp": 1580226060770,
    "baseName": "jhipster",
    "buildTool": "maven",
    "skipUserManagement": true,
    "clientPackageManager": "npm",
    "cacheProvider": "hazelcast",
    "serverPort": "8081",
    "embeddableLaunchScript": false,
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "blueprints": []
  },
  "entities": ["Country"]
}

</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Country {
  name String
}
relationship ManyToMany {
  Country{user(login)} to User
}


</pre>
</details>


##### **Environment and Tools**

openjdk version "12.0.2" 2019-07-16
OpenJDK Runtime Environment AdoptOpenJDK (build 12.0.2+10)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.15.1, JRE 12 Mac OS X amd64-64-Bit Compressed References 20190915_201 (JIT enabled, AOT enabled)
OpenJ9   - 0f66c6431
OMR      - ec782f26
JCL      - 9984ea0dd9 based on jdk-12.0.2+10)

git version 2.25.0

node: v10.18.1

npm: 6.13.4

yarn: 1.21.1

Docker version 19.03.5, build 633a0ea

docker-compose version 1.24.1, build 4667896b

-   [x] Checking this box is mandatory (this is just to show you read everything)

--- a/generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
+++ b/generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
@@ -30,6 +30,13 @@ package <%= packageName %>.web.rest;
             break;
         }
         isUsingMapsId = false;
+    }
+    let manyToManyWithUser = false;
+    for (idx in relationships) {
+        if (relationships[idx].otherEntityName === 'user' && relationships[idx].relationshipType === 'many-to-many') {
+            manyToManyWithUser = true;
+            break;
+        }
     } _%>
 <%_ if (dto !== 'mapstruct' || service === 'no') { _%>
 import <%= packageName %>.domain.<%= asEntity(entityClass) %>;
@@ -166,10 +173,17 @@ public class <%= entityClass %>Resource {
             throw new BadRequestAlertException("A new <%= entityInstance %> cannot already have an ID", ENTITY_NAME, "idexists");
         }
         <%_ if (saveUserSnapshot) { _%>
+            <%_ if (manyToManyWithUser) { _%>
+        if (<%= instanceName %>.getUsers() != null) {
+            // Save users in case they are new and only exists in gateway
+            userRepository.saveAll(<%= instanceName %>.getUsers());
+        }
+            <%_ } else { _%>
         if (<%= instanceName %>.getUser() != null) {
             // Save user in case it's new and only exists in gateway
             userRepository.save(<%= instanceName %>.getUser());
         }
+            <%_ } _%>
         <%_ } _%>
         <%_ if (databaseType === 'cassandra') { _%>
         <%= instanceName %>.setId(UUID.randomUUID());
@@ -196,10 +210,17 @@ public class <%= entityClass %>Resource {
             throw new BadRequestAlertException("Invalid id", ENTITY_NAME, "idnull");
         }
         <%_ if (saveUserSnapshot) { _%>
+            <%_ if (manyToManyWithUser) { _%>
+        if (<%= instanceName %>.getUsers() != null) {
+            // Save users in case they are new and only exists in gateway
+            userRepository.saveAll(<%= instanceName %>.getUsers());
+        }
+            <%_ } else { _%>
         if (<%= instanceName %>.getUser() != null) {
             // Save user in case it's new and only exists in gateway
             userRepository.save(<%= instanceName %>.getUser());
         }
+            <%_ } _%>
         <%_ } _%>
 <%- include('../../common/save_template', {asEntity, asDto, viaService: viaService, returnDirectly: false, isUsingMapsId: false, mapsIdAssoc: mapsIdAssoc}); -%>
         return ResponseEntity.ok()

--- a/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
+++ b/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
@@ -701,10 +701,10 @@ public class <%= entityClass %>ResourceIT <% if (databaseType === 'cassandra') {
     @SuppressWarnings({"unchecked"})
     public void getAll<%= entityClassPlural %>WithEagerRelationshipsIsEnabled() throws Exception {
         <%_ if (service !== 'no') { _%>
-        <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>ServiceMock<% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %>);
+        <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>ServiceMock<% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %><% if (saveUserSnapshot) { %>, userRepository<% } %>);
         when(<%= entityInstance %>ServiceMock.findAllWithEagerRelationships(any())).thenReturn(new PageImpl(new ArrayList<>()));
         <%_ } else { _%>
-        <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>RepositoryMock<% if (dto === 'mapstruct') { %>, <%= entityInstance %>Mapper<% } %><% if (searchEngine === 'elasticsearch') { %>, mock<%= entityClass %>SearchRepository<% } %><% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %>);
+        <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>RepositoryMock<% if (dto === 'mapstruct') { %>, <%= entityInstance %>Mapper<% } %><% if (searchEngine === 'elasticsearch') { %>, mock<%= entityClass %>SearchRepository<% } %><% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %><% if (saveUserSnapshot) { %>, userRepository<% } %>);
         when(<%= entityInstance %>RepositoryMock.findAllWithEagerRelationships(any())).thenReturn(new PageImpl(new ArrayList<>()));
         <%_ } _%>
 
@@ -727,10 +727,10 @@ public class <%= entityClass %>ResourceIT <% if (databaseType === 'cassandra') {
     @SuppressWarnings({"unchecked"})
     public void getAll<%= entityClassPlural %>WithEagerRelationshipsIsNotEnabled() throws Exception {
         <%_ if (service !== 'no') { _%>
-        <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>ServiceMock<% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %>);
+        <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>ServiceMock<% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %><% if (saveUserSnapshot) { %>, userRepository<% } %>);
             when(<%= entityInstance %>ServiceMock.findAllWithEagerRelationships(any())).thenReturn(new PageImpl(new ArrayList<>()));
         <%_ } else { _%>
-        <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>RepositoryMock<% if (dto === 'mapstruct') { %>, <%= entityInstance %>Mapper<% } %><% if (searchEngine === 'elasticsearch') { %>, mock<%= entityClass %>SearchRepository<% } %><% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %>);
+        <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>RepositoryMock<% if (dto === 'mapstruct') { %>, <%= entityInstance %>Mapper<% } %><% if (searchEngine === 'elasticsearch') { %>, mock<%= entityClass %>SearchRepository<% } %><% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %><% if (saveUserSnapshot) { %>, userRepository<% } %>);
             when(<%= entityInstance %>RepositoryMock.findAllWithEagerRelationships(any())).thenReturn(new PageImpl(new ArrayList<>()));
         <%_ } _%>
             MockMvc rest<%= entityClass %>MockMvc = MockMvcBuilders.standaloneSetup(<%= entityInstance %>Resource)
@@ -1036,9 +1036,9 @@ public class <%= entityClass %>ResourceIT <% if (databaseType === 'cassandra') {
         <%= entityInstance %>Repository.save(<%= asEntity(entityInstance) %>);
 
         <%_ if (service !== 'no') { _%>
-        final <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>Service<% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %>);
+        final <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>Service<% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %><% if (saveUserSnapshot) { %>, userRepository<% } %>);
         <%_ } else { _%>
-        final <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>Repository<% if (dto === 'mapstruct') { %>, <%= entityInstance %>Mapper<% } %><% if (searchEngine === 'elasticsearch') { %>, mock<%= entityClass %>SearchRepository<% } %><% if (reactiveRepositories) { %>, <%= entityInstance %>ReactiveRepository<% } %><% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %>);
+        final <%= entityClass %>Resource <%= entityInstance %>Resource = new <%= entityClass %>Resource(<%= entityInstance %>Repository<% if (dto === 'mapstruct') { %>, <%= entityInstance %>Mapper<% } %><% if (searchEngine === 'elasticsearch') { %>, mock<%= entityClass %>SearchRepository<% } %><% if (reactiveRepositories) { %>, <%= entityInstance %>ReactiveRepository<% } %><% if (jpaMetamodelFiltering) { %>, <%= entityInstance %>QueryService<% } %><% if (saveUserSnapshot) { %>, userRepository<% } %>);
         <%_ } _%>
         WebTestClient client = WebTestClient.bindToController(<%= entityInstance %>Resource).build();
 

